src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+            // and doing constant folding without the simplifier undoing the work.
+            if (const Add *add_a = op->a.as<Add>()) {
+                return mutate(Add::make(simplify(Mul::make(add_a->a, op->b)), simplify(Mul::make(add_a->b, op->b))));
+            } else if (const Sub *sub_a = op->a.as<Sub>()) {
+                return mutate(Sub::make(simplify(Mul::make(sub_a->a, op->b)), simplify(Mul::make(sub_a->b, op->b))));
+            }
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        // Rewrite multiplies to shifts if possible.
+        if (op->type.is_int() || op->type.is_uint()) {
+            int pow2 = 0;
+            if (is_const_power_of_two_integer(a, &pow2)) {
+                return mutate(b << cast(UInt(b.type().bits()), pow2));
+            } else if (is_const_power_of_two_integer(b, &pow2)) {
+                return mutate(a << cast(UInt(a.type().bits()), pow2));
+            }
+        }
+
+        // We're applying this to float, which seems OK? float16 * float16 -> float32 is a widening multiply?
+        Expr narrow_a = strip_widening_cast(a);
+        Expr narrow_b = strip_widening_cast(b);
+        if (narrow_a.defined() && narrow_b.defined() &&
+            (narrow_a.type().is_int_or_uint() == narrow_b.type().is_int_or_uint() ||
+             narrow_a.type().is_float() == narrow_b.type().is_float())) {
+            Expr result = widening_mul(narrow_a, narrow_b);
+            if (result.type() != op->type) {
+                result = Cast::make(op->type, result);
+            }
+            return mutate(result);
+        }
+
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return Mul::make(a, b);
+        }
+    }
+
+    Expr visit(const Div *op) override {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        int shift_amount;
+        if (is_const_power_of_two_integer(b, &shift_amount) && op->type.is_int_or_uint()) {
+            return mutate(a >> make_const(UInt(a.type().bits()), shift_amount));
+        }
+
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return Div::make(a, b);
+        }
+    }
+
+    // We don't handle Mod because we don't have any patterns that look for bitwise and vs.
+    // mod.
+
+    template<class MinOrMax>
+    Expr visit_min_or_max(const MinOrMax *op) {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        if (const Cast *cast_a = a.as<Cast>()) {
+            Expr cast_b = lossless_cast(cast_a->value.type(), b);
+            if (cast_a->type.can_represent(cast_a->value.type()) && cast_b.defined()) {
+                // This cast can be moved outside the min.
```

Comment:
Is this a good idea if it's a narrowing cast? Surely we'd rather do the min in the narrower type?

Oh, the can_represent means that it's a widening cast of some flavor (or uint8 -> float or something). Please update comment.